core/shared/src/main/scala/zio/ZIO.scala

```diff
@@ -5427,7 +5427,7 @@ object ZIO extends ZIOCompanionPlatformSpecific {
   }
 
   final class ServiceWithPartiallyApplied[Service](private val dummy: Boolean = true) extends AnyVal {
-    def apply[E, A](f: Service => ZIO[Has[Service], E, A])(implicit tag: Tag[Service]): ZIO[Has[Service], E, A] =
+    def apply[R <: Has[Service], E, A](f: Service => ZIO[R, E, A])(implicit tag: Tag[Service]): ZIO[R with Has[Service], E, A] =
```

Comment:
Will this work:

```suggestion
    def apply[R, E, A](f: Service => ZIO[R, E, A])(implicit tag: Tag[Service]): ZIO[R, E, A] =
```

Reply:
That was my original formulation. 😃

The issue there is that it leads to a lot of spurious warnings about `R` being inferred as `Any`.

Reply:
Or to be more precise there are a couple of permutations:

```scala
    def apply[R, E, A](f: Service => ZIO[R, E, A])(implicit
      tag: Tag[Service]
    ): ZIO[R with Has[Service], E, A] =
```

This one leads to spurious warnings about a type being inferred as `Any`.

```scala
    def apply[R <: Has[Service], E, A](f: Service => ZIO[R, E, A])(implicit
      tag: Tag[Service]
    ): ZIO[R, E, A]
```

This one leads to numerous warnings of missing parameter type for the effectual function.